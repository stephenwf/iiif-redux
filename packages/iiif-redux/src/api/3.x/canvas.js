import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationPages,
  getAllCanvases,
  getAllContentResources,
  getAllExternalResources,
  getAllResources,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';

const canvas = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getHeight
   * - getWidth
   * - getDuration
   * - getBehavior
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getHeight = createSelector(selector, technical.getHeight);

  const getWidth = createSelector(selector, technical.getWidth);

  const getDuration = createSelector(selector, technical.getDuration);

  const getBehavior = createSelector(selector, technical.getBehavior);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getPosterCanvas
   * - getRequiredStatement
   * - getRights
   * - getNavDate
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getPosterCanvasId = createSelector(
    selector,
    descriptive.getPosterCanvas
  );

  const getPosterCanvas = createSelector(
    getPosterCanvasId,
    getAllCanvases,
    mapById
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  const getNavDate = createSelector(selector, descriptive.getNavDate);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllExternalResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllExternalResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllExternalResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  /**
   * Structural Properties
   *
   * - getItems
   * - getStructures
   * - getAnnotations
   */
  const getItemIds = createSelector(selector, structural.getItems);
  const getItems = createSelector(getItemIds, getAllCanvases, mapAllById);

  const getPaintingAnnotationIds = getItemIds;
  const getPaintingAnnotations = getItems;

  const getAnnotationIds = createSelector(selector, structural.getAnnotations);
  const getAnnotations = createSelector(
    getAnnotationIds,
    getAllAnnotationPages,
    mapAllById
  );

  return {
    getId,
    getType,
    getHeight,
    getWidth,
    getDuration,
    getBehavior,
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getPosterCanvasId,
    getPosterCanvas,
    getRequiredStatement,
    getRights,
    getNavDate,
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
    getItemIds,
    getItems,
    getPaintingAnnotationIds,
    getPaintingAnnotations,
    getAnnotationIds,
    getAnnotations,
  };
});

export default canvas;
