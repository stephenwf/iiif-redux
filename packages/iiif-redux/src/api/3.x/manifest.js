import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationPages,
  getAllCanvases,
  getAllCanvasReferences,
  getAllContentResources,
  getAllRanges,
  getAllResources,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';
import { standardFieldMappingFactory } from '../../../es/utility/new/standardFieldMappingFactory';

const manifest = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingDirection
   * - getBehavior
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getBehavior = createSelector(selector, technical.getBehavior);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getPosterCanvas
   * - getRequiredStatement
   * - getRights
   * - getNavDate
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getPosterCanvasId = createSelector(
    selector,
    descriptive.getPosterCanvas
  );

  const getPosterCanvas = createSelector(
    getPosterCanvasId,
    getAllCanvases,
    mapById
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  const getNavDate = createSelector(selector, descriptive.getNavDate);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   * - getStart
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllContentResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllContentResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllContentResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  const getStartId = createSelector(selector, linking.getStart);
  const getStart = createSelector(
    getStartId,
    getAllCanvases,
    getAllCanvasReferences,
    (start, allCanvases, allCanvasReferences) => {
      if (!start) {
        return null;
      }
      if (start.schema === 'canvas') {
        return allCanvases[start.id];
      }
      return allCanvasReferences[start.id];
    }
  );

  /**
   * Structural Properties
   *
   * - getItems
   * - getStructures
   * - getAnnotations
   */
  const getItemIds = createSelector(selector, structural.getItems);
  const getItems = createSelector(getItemIds, getAllCanvases, mapAllById);

  const getCanvasIds = getItemIds;
  const getCanvases = getItems;

  const getStructureIds = createSelector(selector, structural.getStructures);
  const getStructures = createSelector(
    getStructureIds,
    getAllRanges,
    mapAllById
  );

  const getAnnotationIds = createSelector(selector, structural.getAnnotations);
  const getAnnotations = createSelector(
    getAnnotationIds,
    getAllAnnotationPages,
    mapAllById
  );

  return {
    getId,
    getType,
    getViewingDirection,
    getBehavior,
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getPosterCanvasId,
    getPosterCanvas,
    getRequiredStatement,
    getRights,
    getNavDate,
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
    getStartId,
    getStart,
    getItemIds,
    getItems,
    getCanvasIds,
    getCanvases,
    getStructureIds,
    getStructures,
    getAnnotationIds,
    getAnnotations,
  };
});

export const mappings = standardFieldMappingFactory(manifest);

export default manifest;
