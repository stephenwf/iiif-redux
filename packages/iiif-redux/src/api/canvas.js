import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import validUrl from 'valid-url';
import * as technical from './iiif-technical';
import * as descriptive from './iiif-descriptive';
import * as linking from './iiif-linking';
import * as structural from './iiif-structural';
import {
  getAllAnnotationLists,
  getAllAnnotations,
  getAllExternalResources,
  getAllImages,
  getAllLayers,
  getAllServices,
} from './all';
import { isImageService } from '../constants/services';

const createCanvasSelectors = memoize(selector => {
  /**************************************************
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingHint
   * - getHeight
   * - getWidth
   **************************************************/
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingHint = createSelector(selector, canvas => {
    const viewingHint = technical.getViewingHint(canvas);
    switch (viewingHint) {
      case 'non-paged':
      case 'facing-pages':
        return viewingHint;
      default:
        return validUrl.isWebUri(viewingHint) ? viewingHint : null;
    }
  });

  const getHeight = createSelector(selector, technical.getHeight);

  const getWidth = createSelector(selector, technical.getWidth);

  /**************************************************
   * Descriptive properties
   *
   * - getLabel
   * - getMetadata
   * - getDescription
   * - getThumbnail
   * - getAttribution
   * - getLicense
   * - getLogo
   **************************************************/
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getDescription = createSelector(selector, descriptive.getDescription);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getAttribution = createSelector(selector, descriptive.getAttribution);

  const getLogo = createSelector(selector, descriptive.getLogo);

  const getLicense = createSelector(selector, descriptive.getLicense);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);
  const getThumbnail = createSelector(
    getThumbnailId,
    getAllImages,
    (thumbnailId, allImages) => allImages[thumbnailId] || thumbnailId
  );

  /**************************************************
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getRelated
   * - getRendering
   * - getWithin
   **************************************************/
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllExternalResources,
    (seeAlsoIds, allExternalResources) =>
      seeAlsoIds.map(
        seeAlsoId =>
          allExternalResources[seeAlsoId] || {
            '@id': seeAlsoId,
            label: 'unknown',
          }
      )
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(
    getServiceIds,
    getAllServices,
    (serviceIds, allServices) =>
      serviceIds.map(serviceId => allServices[serviceId])
  );

  const getRelatedIds = createSelector(selector, linking.getRelated);
  const getRelated = createSelector(
    getRelatedIds,
    getAllExternalResources,
    (relatedIds, allExternalResources) =>
      relatedIds.map(relatedId => allExternalResources[relatedId])
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllExternalResources,
    (renderingIds, allExternalResources) =>
      renderingIds.map(renderingId => allExternalResources[renderingId])
  );

  const getWithinIds = createSelector(selector, linking.getWithin);
  const getWithin = createSelector(
    getWithinIds,
    getAllLayers,
    getAllExternalResources,
    (withinIds, allLayers, allExternalResources) =>
      withinIds.map(
        withinId => allLayers[withinId] || allExternalResources[withinId]
      )
  );

  /**************************************************
   * Structural properties
   *
   * - getOtherContent
   * - getImages
   **************************************************/
  const getOtherContentIds = createSelector(
    selector,
    structural.getOtherContent
  );
  const getOtherContent = createSelector(
    getOtherContentIds,
    getAllExternalResources,
    getAllAnnotationLists,
    (otherContentIds, allExternalResources, allAnnotationLists) =>
      otherContentIds.map(
        otherContentId =>
          allAnnotationLists[otherContentId] ||
          allExternalResources[otherContentId] || {
            '@id': otherContentId,
            label: 'unknown',
          }
      )
  );

  const getImageIds = createSelector(selector, structural.getImages);
  const getImages = createSelector(
    getImageIds,
    getAllAnnotations,
    (imageIds, allImages) => imageIds.map(imageId => allImages[imageId])
  );

  /**************************************************
   * Algorithms
   *
   * - getImageService
   **************************************************/
  const getImageService = createSelector(
    getImages,
    getAllServices,
    getAllImages,
    (images, allServices, allImages) =>
      // The basic path is images[x].resource.service
      images.reduce((result, image) => {
        if (result) return result;
        const resource =
          image.resource.schema === 'imageResource'
            ? allImages[image.resource.id]
            : { service: [] };

        return resource.service.reduce((innerResult, serviceId) => {
          if (innerResult) return innerResult;
          const service = allServices[serviceId];
          if (isImageService(service.profile)) {
            return service;
          }
        }, null);
      }, null)
  );

  return {
    // Technical
    getId,
    getType,
    getViewingHint,
    getHeight,
    getWidth,
    // Descriptive
    getLabel,
    getMetadata,
    getDescription,
    getThumbnailId,
    getThumbnail,
    getAttribution,
    getLicense,
    getLogo,
    // Linking
    getWithinIds,
    getWithin,
    getRenderingIds,
    getRendering,
    getRelatedIds,
    getRelated,
    getServiceIds,
    getService,
    getSeeAlsoIds,
    getSeeAlso,
    // Structural
    getOtherContentIds,
    getOtherContent,
    getImageIds,
    getImages,
    // Algorithms.
    getImageService,
  };
});

export default createCanvasSelectors;
