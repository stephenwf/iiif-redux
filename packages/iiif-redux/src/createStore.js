import { createStore, applyMiddleware, combineReducers, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import thunk from 'redux-thunk';
import * as reducers from './reducers';
import sagas from './sagas';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default function createCustomStore(
  customReducers = {},
  extraMiddleware = [],
  customSagas = [],
  defaultState = {}
) {
  const sagaMiddleware = createSagaMiddleware();

  const store = createStore(
    combineReducers(reducers),
    defaultState,
    composeEnhancers(applyMiddleware(sagaMiddleware, thunk, ...extraMiddleware))
  );

  sagas.map(saga => sagaMiddleware.run(saga));
  // customSagas.map(saga => sagaMiddleware.run(saga));

  return store;
}
