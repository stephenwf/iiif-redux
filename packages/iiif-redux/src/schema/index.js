import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
