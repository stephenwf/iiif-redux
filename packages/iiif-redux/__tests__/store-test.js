import { createStructuredSelector, createSelector } from 'reselect';
import createStore from '../src/createStore';
import bridges from './fixtures/bridges';
import {
  iiifResourceRequest,
  iiifResourceRequestUnknown,
} from '../src/spaces/iiif-resource';
import { collection, manifest } from '../src/schema/presentation2';
import { waitForRequest } from '../test-utils';
import { manifestByIdSelector } from '../src/api/manifest';
import { sequenceByIdSelector } from '../src/api/sequence';
import { canvasByIdSelector } from '../src/api/canvas';

describe('store', () => {
  global.fetch = require('jest-fetch-mock');
  const RealDate = Date;

  function mockDate(isoDate) {
    // noinspection JSUnresolvedVariable
    global.Date = class extends RealDate {
      // noinspection JSAnnotator
      constructor(...theArgs) {
        if (theArgs.length) {
          return new RealDate(...theArgs);
        }
        return new RealDate(isoDate);
      }

      static now() {
        return new RealDate(isoDate).getTime();
      }
    };
  }
  afterEach(() => {
    // noinspection JSUnresolvedVariable
    global.Date = RealDate;
  });

  it('should have default state', () => {
    const store = createStore();
    expect(store.getState()).toMatchSnapshot();
  });

  it('should throw an error if actions are not passed through', () => {
    mockDate('2017-11-25T00:00:00z');

    const store = createStore();
    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS'],
        manifest
      )
    );

    const state = store.getState();
    expect(state).toMatchSnapshot();
  });

  it('should set error in state when request errors', async () => {
    const store = createStore();
    mockDate('2017-11-25T00:00:00z');
    fetch.mockRejectOnce('Excepted error found');
    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );
    await new Promise(resolve => store.subscribe(resolve));
    const state = store.getState();
    expect(state).toMatchSnapshot();
  });

  it('should set error in state when malformed url is given', () => {
    const store = createStore();
    mockDate('2017-11-25T00:00:00z');
    store.dispatch(
      iiifResourceRequest(
        'not-real-url',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );

    const state = store.getState();

    expect(state).toMatchSnapshot();
  });

  it('should import a manifest', async () => {
    const store = createStore();
    fetch.mockResponseOnce(JSON.stringify(bridges));
    mockDate('2017-11-25T00:00:00z');

    const whenRequestFinishes = waitForRequest(
      store,
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );

    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );
    const state = store.getState();

    const manifestState =
      state.dereferenced[
        'https://view.nls.uk/manifest/7446/74464117/manifest.json'
      ];

    expect(manifestState.loading).toEqual(true);
    expect(manifestState.resourceId).toEqual(
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );
    expect(manifestState.ttl).toEqual(600);

    await whenRequestFinishes;

    const secondState = store.getState();
    expect(secondState).toMatchSnapshot('second state');

    // Try fetching as unknown.
    fetch.mockResponseOnce(JSON.stringify(bridges));
    const whenRequestFinishes2 = waitForRequest(
      store,
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );

    store.dispatch(
      iiifResourceRequestUnknown(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json'
      )
    );

    await whenRequestFinishes2;

    expect(
      store.getState().dereferenced[
        'https://view.nls.uk/manifest/7446/74464117/manifest.json'
      ]
    ).toEqual({
      loading: false,
      resourceId: 'https://view.nls.uk/manifest/7446/74464117/manifest.json',
      ttl: 600,
    });
  });

  it('should only make 1 http request per resource by default', async () => {
    const store = createStore();
    fetch.mockResponseOnce(JSON.stringify(bridges));

    const whenRequestFinishes = waitForRequest(
      store,
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );

    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );
    const state = store.getState();

    const manifestState =
      state.dereferenced[
        'https://view.nls.uk/manifest/7446/74464117/manifest.json'
      ];

    expect(manifestState.loading).toEqual(true);
    expect(manifestState.resourceId).toEqual(
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );
    expect(manifestState.ttl).toEqual(600);

    await whenRequestFinishes;

    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );

    const state2 = store.getState();
    expect(
      state2.dereferenced[
        'https://view.nls.uk/manifest/7446/74464117/manifest.json'
      ].loading
    ).toEqual(false);
  });

  it('should only correct incorrect IDs when requesting resource.', async () => {
    const store = createStore();
    fetch.mockResponseOnce(
      JSON.stringify({ ...bridges, '@id': 'https://INCORRECT_FIX_ME/' })
    );

    const whenRequestFinishes = waitForRequest(
      store,
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );

    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );
    const state = store.getState();

    const manifestState =
      state.dereferenced[
        'https://view.nls.uk/manifest/7446/74464117/manifest.json'
      ];

    expect(manifestState.loading).toEqual(true);
    expect(manifestState.resourceId).toEqual(
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );
    expect(manifestState.ttl).toEqual(600);

    await whenRequestFinishes;

    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );

    const state2 = store.getState();
    expect(
      state2.dereferenced[
        'https://view.nls.uk/manifest/7446/74464117/manifest.json'
      ].loading
    ).toEqual(false);
  });

  it('should create state that works with selectors', async () => {
    const store = createStore({}, [], [], {
      routing: {
        currentManifest:
          'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        currentSequence:
          'https://view.nls.uk/manifest/7446/74464117/canvas/default',
        currentCanvas: 'https://view.nls.uk/iiif/7446/74464117/canvas/1',
      },
    });
    fetch.mockResponseOnce(JSON.stringify(bridges));

    const whenRequestFinishes = waitForRequest(
      store,
      'https://view.nls.uk/manifest/7446/74464117/manifest.json'
    );

    store.dispatch(
      iiifResourceRequest(
        'https://view.nls.uk/manifest/7446/74464117/manifest.json',
        ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
        manifest
      )
    );

    await whenRequestFinishes;

    const state = store.getState();

    // Assert manifest.
    const manifestLabel = manifestByIdSelector(api => api.getLabel, {
      getId: () => 'https://view.nls.uk/manifest/7446/74464117/manifest.json',
    });
    expect(manifestLabel(state)).toMatchSnapshot();

    // Assert sequence.
    const sequenceLabel = sequenceByIdSelector(api => api.getLabel, {
      getId: () => 'https://view.nls.uk/manifest/7446/74464117/canvas/default',
    });

    expect(sequenceLabel(state)).toMatchSnapshot();

    // Assert canvas.
    const canvasState = canvasByIdSelector(
      api => ({
        label: api.getLabel,
        thumbnail: api.getThumbnailId,
        imageService: api.getImageService,
      }),
      {
        getId: () => 'https://view.nls.uk/iiif/7446/74464117/canvas/1',
      }
    );

    expect(canvasState(state)).toMatchSnapshot();
  });

  it('should preload normalized entities in collections', async () => {
    const collectionJson = {
      '@id': 'http://iiif.com/collection-1.json',
      '@type': 'sc:Collection',
      label: 'Collection label 1',
      members: [
        {
          '@id': 'http://iiif.com/collection-2.json',
          '@type': 'sc:Collection',
          label: 'Collection label 2',
        },
        {
          '@id': 'http://iiif.com/collection-3.json',
          '@type': 'sc:Collection',
          label: 'Collection label 3',
        },
      ],
      manifests: [
        {
          '@id': 'http://iiif.com/manifest-1.json',
          '@type': 'sc:Manifest',
          label: 'Manifest label 1',
        },
        {
          '@id': 'http://iiif.com/manifest-2.json',
          '@type': 'sc:Manifest',
          label: 'Manifest label 2',
        },
      ],
    };

    const store = createStore();
    fetch.mockResponseOnce(JSON.stringify(collectionJson));
    mockDate('2017-11-25T00:00:00z');

    const whenRequestFinishes = waitForRequest(
      store,
      'http://iiif.com/collection-1.json'
    );

    store.dispatch(
      iiifResourceRequest(
        'http://iiif.com/collection-1.json',
        ['COLLECTION_REQUEST', 'COLLECTION_SUCCESS', 'COLLECTION_ERROR'],
        collection
      )
    );

    await whenRequestFinishes;

    const state = store.getState();

    expect(state.resources).toMatchSnapshot();
  });

  describe('iiifResourceRequestUnknown', () => {
    test('basic resource', () => {
      expect(
        iiifResourceRequestUnknown('https://iiif.com/collection-1.json')
      ).toMatchSnapshot();
    });

    test('http request', async () => {
      const store = createStore();
      fetch.mockResponseOnce(
        JSON.stringify({
          '@id': 'http://iiif.com/collection-1.json',
          '@type': 'sc:Collection',
          label: 'Collection label 1',
        })
      );

      fetch.mockResponseOnce(
        JSON.stringify({
          '@id': 'http://iiif.com/collection-1.json',
          '@type': 'sc:Collection',
          label: 'Collection label 1',
        })
      );

      const whenRequestFinishes = waitForRequest(
        store,
        'http://iiif.com/collection-1.json'
      );

      store.dispatch(
        iiifResourceRequestUnknown('http://iiif.com/collection-1.json')
      );

      await whenRequestFinishes;

      const state = store.getState();

      expect(state.dereferenced.error).toEqual();

      expect(state.resources).toMatchSnapshot();
    });

    test('invalid URL', async () => {
      const store = createStore();

      store.dispatch(iiifResourceRequestUnknown('NOT-REAL-URL'));
      const state = store.getState();

      expect(state.dereferenced['NOT-REAL-URL'].error).toEqual(
        'Resource is not a valid URL.'
      );
    });

    test('unknown type', async () => {
      const store = createStore();
      fetch.mockResponseOnce(
        JSON.stringify({
          '@id': 'http://iiif.com/NOT-REAL-1.json',
          '@type': 'NOT REAL',
          label: 'Collection label 1',
        })
      );
      const whenRequestFinishes = waitForRequest(
        store,
        'http://iiif.com/NOT-REAL-1.json'
      );
      store.dispatch(
        iiifResourceRequestUnknown('http://iiif.com/NOT-REAL-1.json')
      );

      await whenRequestFinishes;

      const state = store.getState();

      expect(
        state.dereferenced['http://iiif.com/NOT-REAL-1.json'].error
      ).toEqual('Unknown Presentation 2 resource type');
    });

    test('unknown mapping', async () => {
      const store = createStore();
      fetch.mockResponseOnce(
        JSON.stringify({
          '@id': 'http://iiif.com/NOT-REAL-1.json',
          '@type': 'sc:Collection',
          label: 'Collection label 1',
        })
      );
      const whenRequestFinishes = waitForRequest(
        store,
        'http://iiif.com/NOT-REAL-1.json'
      );
      store.dispatch(
        iiifResourceRequestUnknown(
          'http://iiif.com/NOT-REAL-1.json',
          {},
          {
            canvas: () => null, // noop for test
            manifest: () => null, // noop for test
          }
        )
      );

      await whenRequestFinishes;

      const state = store.getState();

      expect(
        state.dereferenced['http://iiif.com/NOT-REAL-1.json'].error
      ).toEqual(
        'Resource type is not in configured mappings (canvas, manifest)'
      );
    });
  });
});
