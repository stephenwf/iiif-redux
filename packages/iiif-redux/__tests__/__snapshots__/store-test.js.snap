// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store iiifResourceRequestUnknown basic resource 1`] = `
Object {
  "payload": Object {
    "dereferenceIfMissing": false,
    "forceFresh": undefined,
    "mappings": Object {
      "collection": Array [
        "COLLECTION_REQUEST",
        "COLLECTION_SUCCESS",
        "COLLECTION_ERROR",
      ],
      "manifest": Array [
        "MANIFEST_REQUEST",
        "MANIFEST_SUCCESS",
        "MANIFEST_ERROR",
      ],
    },
    "resourceId": "https://iiif.com/collection-1.json",
    "ttl": 600,
  },
  "type": "IIIF_RESOURCE_REQUEST_UNKNOWN",
}
`;

exports[`store iiifResourceRequestUnknown http request 1`] = `
Object {
  "annotationLists": Object {},
  "annotations": Object {},
  "canvases": Object {},
  "collections": Object {
    "http://iiif.com/collection-1.json": Object {
      "@id": "http://iiif.com/collection-1.json",
      "@type": "sc:Collection",
      "label": Array [
        Object {
          "@language": "en",
          "@value": "Collection label 1",
        },
      ],
    },
  },
  "externalResources": Object {},
  "imageResources": Object {},
  "layers": Object {},
  "manifests": Object {},
  "ranges": Object {},
  "sequences": Object {},
}
`;
